In [1]:
import numpy as np
import pandas as pd

In [37]:
import time

In [2]:
import scipy.io

In [3]:
from anomaly.base import AnomalyDetector

In [4]:
from river import metrics
from river.stream import iter_array

In [5]:
from models.xstream import xStream

In [7]:
#ROCAUC scikit-learn
from pysad.evaluation import AUROCMetric

In [6]:
np.random.seed(61)

In [17]:
model = xStream()
metric = AUROCMetric()

# Arrythmia (outliers 15%) nb_points: 452

In [51]:
data = scipy.io.loadmat('/Users/Sophie/Library/CloudStorage/OneDrive-Personnel/IP/Data Stream Processing/PySAD_to_River/data/arrhythmia.mat') 

In [52]:
X = data['X']
y = data['y']
y_flat = data['y'].flatten()

In [ ]:
start = time.time()
for xi, yi in iter_array(X,y_flat):
    anomaly_score = model.score_one(xi)
    model = model.learn_one(xi)
    metric.update(yi, anomaly_score)  # Update the AUROC metric.
#    print(anomaly_score)
# Output the resulting AUROCMetric.
run_time = time.time() - start
print(f"Area under ROC metric is {metric.get()}.")
print(run_time)

Area under ROC metric is 0.7061155597425027.

# Optdigits (outliers 3%) nb_points: 5216

In [28]:
data = scipy.io.loadmat('/Users/Sophie/Library/CloudStorage/OneDrive-Personnel/IP/Data Stream Processing/PySAD_to_River/data/optdigits.mat') 

In [29]:
X = data['X']
y = data['y']
y_flat = data['y'].flatten()

In [30]:
start = time.time()
for xi, yi in iter_array(X,y_flat):
    anomaly_score = model.score_one(xi)
    model = model.learn_one(xi)
    metric.update(yi, anomaly_score)  # Update the AUROC metric.
#    print(anomaly_score)
# Output the resulting AUROCMetric.
run_time = time.time() - start
print(f"Area under ROC metric is {metric.get()}.")
print(run_time)

Area under ROC metric is 0.6683226856316297.


Area under ROC metric is 0.6683226856316297.

# Wine (outliers 7.7%) nb_points: 129

In [25]:
data = scipy.io.loadmat('/Users/Sophie/Library/CloudStorage/OneDrive-Personnel/IP/Data Stream Processing/PySAD_to_River/data/wine.mat') 

In [26]:
X = data['X']
y = data['y']
y_flat = data['y'].flatten()

In [27]:
start = time.time()
for xi, yi in iter_array(X,y_flat):
    anomaly_score = model.score_one(xi)
    model = model.learn_one(xi)
    metric.update(yi, anomaly_score)  # Update the AUROC metric.
#    print(anomaly_score)
# Output the resulting AUROCMetric.
run_time = time.time() - start
print(f"Area under ROC metric is {metric.get()}.")
print(run_time)

Area under ROC metric is 0.5781512605042017.


Area under ROC metric is 0.5781512605042017.

# Pendigits (outliers 2.27%) nb_points: 6870

In [31]:
data = scipy.io.loadmat('/Users/Sophie/Library/CloudStorage/OneDrive-Personnel/IP/Data Stream Processing/PySAD_to_River/data/pendigits.mat') 

In [32]:
X = data['X']
y = data['y']
y_flat = data['y'].flatten()

In [33]:
start = time.time()
for xi, yi in iter_array(X,y_flat):
    anomaly_score = model.score_one(xi)
    model = model.learn_one(xi)
    metric.update(yi, anomaly_score)  # Update the AUROC metric.
#    print(anomaly_score)
# Output the resulting AUROCMetric.
run_time = time.time() - start
print(f"Area under ROC metric is {metric.get()}.")
print(run_time)

Area under ROC metric is 0.7631021009105116.


Area under ROC metric is 0.7631021009105116.

# Mnist (outliers 9.2%) nb_points: 7603

In [34]:
data = scipy.io.loadmat('/Users/Sophie/Library/CloudStorage/OneDrive-Personnel/IP/Data Stream Processing/PySAD_to_River/data/mnist.mat') 

In [35]:
X = data['X']
y = data['y']
y_flat = data['y'].flatten()

In [ ]:
start = time.time()
for xi, yi in iter_array(X,y_flat):
    anomaly_score = model.score_one(xi)
    model = model.learn_one(xi)
    metric.update(yi, anomaly_score)  # Update the AUROC metric.
#    print(anomaly_score)
# Output the resulting AUROCMetric.
run_time = time.time() - start
print(f"Area under ROC metric is {metric.get()}.")
print(run_time)

Area under ROC metric is 0.50121249037839.

# Annthyroid (outliers 7.42%) nb_points: 7200

In [39]:
data = scipy.io.loadmat('/Users/Sophie/Library/CloudStorage/OneDrive-Personnel/IP/Data Stream Processing/PySAD_to_River/data/annthyroid.mat') 

In [40]:
X = data['X']
y = data['y']
y_flat = data['y'].flatten()

In [41]:
start = time.time()
for xi, yi in iter_array(X,y_flat):
    anomaly_score = model.score_one(xi)
    model = model.learn_one(xi)
    metric.update(yi, anomaly_score)  # Update the AUROC metric.
#    print(anomaly_score)
# Output the resulting AUROCMetric.
run_time = time.time() - start
print(f"Area under ROC metric is {metric.get()}.")
print(run_time)

Area under ROC metric is 0.539149900950972.
798.5197999477386


Area under ROC metric is 0.539149900950972.

Execution time: 798.5197999477386, 13min

# Vertebral (outliers 12.5%) nb_points: 240

In [42]:
data = scipy.io.loadmat('/Users/Sophie/Library/CloudStorage/OneDrive-Personnel/IP/Data Stream Processing/PySAD_to_River/data/vertebral.mat') 

In [43]:
X = data['X']
y = data['y']
y_flat = data['y'].flatten()

In [44]:
start = time.time()
for xi, yi in iter_array(X,y_flat):
    anomaly_score = model.score_one(xi)
    model = model.learn_one(xi)
    metric.update(yi, anomaly_score)  # Update the AUROC metric.
#    print(anomaly_score)
# Output the resulting AUROCMetric.
run_time = time.time() - start
print(f"Area under ROC metric is {metric.get()}.")
print(run_time)

Area under ROC metric is 0.5427017019938262.
28.776970148086548


Area under ROC metric is 0.5427017019938262.

Execution time: 28.776970148086548, 28s

# Satellite (outliers 32%) nb_points: 6435

In [45]:
data = scipy.io.loadmat('/Users/Sophie/Library/CloudStorage/OneDrive-Personnel/IP/Data Stream Processing/PySAD_to_River/data/satellite.mat') 

In [46]:
X = data['X']
y = data['y']
y_flat = data['y'].flatten()

In [47]:
start = time.time()
for xi, yi in iter_array(X,y_flat):
    anomaly_score = model.score_one(xi)
    model = model.learn_one(xi)
    metric.update(yi, anomaly_score)  # Update the AUROC metric.
#    print(anomaly_score)
# Output the resulting AUROCMetric.
run_time = time.time() - start
print(f"Area under ROC metric is {metric.get()}.")
print(run_time)

Area under ROC metric is 0.670962463857515.
750.2115750312805


Area under ROC metric is 0.670962463857515.

Execution time: 750.2115750312805, 13min

# BreastW (outliers 35%) nb_points: 683

In [48]:
data = scipy.io.loadmat('/Users/Sophie/Library/CloudStorage/OneDrive-Personnel/IP/Data Stream Processing/PySAD_to_River/data/breastw.mat') 

In [49]:
X = data['X']
y = data['y']
y_flat = data['y'].flatten()

In [50]:
start = time.time()
for xi, yi in iter_array(X,y_flat):
    anomaly_score = model.score_one(xi)
    model = model.learn_one(xi)
    metric.update(yi, anomaly_score)  # Update the AUROC metric.
#    print(anomaly_score)
# Output the resulting AUROCMetric.
run_time = time.time() - start
print(f"Area under ROC metric is {metric.get()}.")
print(run_time)

Area under ROC metric is 0.6735313667003362.
72.90796303749084


Area under ROC metric is 0.6735313667003362.

Execution time: 72.90796303749084, 1min